tests/scenario/test_extension_providers.py

```diff
@@ -0,0 +1,203 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+from scenario import State
+
+from charm import IstioCoreCharm, flatten_config
+
+#############################################
+# Tests for flatten_config
+#############################################
+
+
+def test_flatten_config_with_flat_dict():
+    """Verify that a flat dictionary is correctly flattened when a prefix is provided."""
+    data = {"a": "value1", "b": "value2"}
+    prefix = "provider[0]"
+    expected = {
+        "provider[0].a": "value1",
+        "provider[0].b": "value2",
+    }
+    result = flatten_config(data, prefix)
+    assert result == expected
+
+
+def test_flatten_config_with_nested_dict():
+    """Verify that nested dictionaries are recursively flattened with the provided prefix."""
+    data = {"a": {"b": 1, "c": {"d": "x"}}, "e": 2}
+    prefix = "prefix"
+    expected = {
+        "prefix.a.b": 1,
+        "prefix.a.c.d": "x",
+        "prefix.e": 2,
+    }
+    result = flatten_config(data, prefix)
+    assert result == expected
```

Comment:
Any reason why we don't use `pytest.mark.parametrize` here?  This looks like 3 parametrized tests rather than 13 independent tests

Using the top two as an example:
```suggestion
@pytest.mark.parametrize(
  "data, prefix, expected",
  [
    # flat dict
    (
      {"a": "value1", "b": "value2"}, 
      "provider[0]",
      {
        "provider[0].a": "value1",
        "provider[0].b": "value2",
      },
    ),
    # nested dict
    (
      {"a": {"b": 1, "c": {"d": "x"}}, "e": 2}, 
      "prefix",
      {
        "prefix.a.b": 1,
        "prefix.a.c.d": "x",
        "prefix.e": 2,
      },
    ),
    # ...
  ]
)    
def test_flatten_config(data, prefix, expected):
    result = flatten_config(data, prefix)
    assert result == expected

```

the scenario tests below are the same, too

Reply:
I personally try to refrain from using `pytest.mark.parametrize` just because I like how separate small tests are more readable and have clearer descriptions and code reduction isn't really adding anything to overall performance, I changed them all tho to use `parametrize` to stay consistent with our charms  :angel: